torch/csrc/jit/python_compiled_function.cpp

```diff
@@ -97,15 +97,30 @@ struct CompiledFunction {
       return fn->apply(inputs);
     }
 
-    PyObject* add_trace(PyObject *args, variable_list inputs) {
+    PyObject* add_trace(PyObject *args, ParsedArgs input_info) {
       JIT_ASSERT(!is_ready_);
       // Start tracing
       AutoGradMode grad_mode(grad_enabled_);
       auto num_stages = grad_enabled_ ? fn_.nderivs_ + 1 : 1;
-      auto trace = tracer::enter(fmap<TraceInput>(inputs), num_stages);
+      auto enter_info = tracer::enter(fmap<TraceInput>(input_info.vars), num_stages);
+      auto & trace = enter_info.first;
+      auto & new_vars = enter_info.second;
+
+      // Enter returns us a new list of Variables to use as inputs, so handle that.
+      std::size_t num_all_inputs = input_info.vars.size();
+      std::size_t num_captured = fn_.captured_vars_.size();
+      // Check that no captured Variables were replaced by enter. It's hard to handle that.
```

Comment:
It is guaranteed that this does not happen (it is an assertion so I assume yes).

Reply:
It's not guaranteed, it's more of a check. I forgot we have `JIT_EXPECT`. I'll change it to use that